larray_editor/arrayadapter.py

```diff
@@ -3,188 +3,557 @@
+            ax.set_xticks(xticks)
+            ax.set_xticklabels(xticklabels)
+
+            if raw_data.shape[1] != 1 and ylabels != [[]]:
+                # set legend
+                # box = ax.get_position()
+                # ax.set_position([box.x0, box.y0, box.width * 0.85, box.height])
+                # ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))
+                ax.legend()
+
+            return figure
+
+
+@register_adapter(np.ndarray)
+@register_adapter(la.LArray)
+class LArrayDataAdapter(AbstractAdapter):
+    def __init__(self, data, changes, bg_value):
+        AbstractAdapter.__init__(self, data=data, changes=changes, bg_value=bg_value)
+        self.ndim = data.ndim
+        self.size = data.size
+        self.dtype = data.dtype
+
+    def prepare_data(self, data):
+        return la.aslarray(data)
+
+    def prepare_bg_value(self, bg_value):
+        return la.aslarray(bg_value) if bg_value is not None else None
+
+    def filter_data(self, data, filter):
+        if data is None:
+            return data
+        assert isinstance(data, la.LArray)
+        if filter is None:
+            return data
+        else:
+            assert isinstance(filter, dict)
+            data = data[filter]
+            return la.aslarray(data) if np.isscalar(data) else data
+
+    def get_axes(self, data):
+        assert isinstance(data, la.LArray)
+        axes = data.axes
+        # test data.size == 0 is required in case an instance built as LArray([]) is passed
+        # test len(axes) == 0 is required when a user filters until to get a scalar
+        if data.size == 0 or len(axes) == 0:
+            return []
+        else:
+            return [Axis(axes.axis_id(axis), name, axis.labels) for axis, name in zip(axes, axes.display_names)]
+
+    def _get_raw_data(self, data):
+        assert isinstance(data, la.LArray)
+        return data.data
+
+    def _get_bg_value(self, bg_value):
+        if bg_value is not None:
+            assert isinstance(bg_value, la.LArray)
+            return bg_value.data
+        else:
+            return bg_value
+
+    # TODO: update this method the day LArray objects will also handle MultiIndex-like axes.
+    def from_selection(self, raw_data, axes_names, vlabels, hlabels):
+        axes = []
+        # combine the N-1 first axes
+        if len(axes_names) > 1:
+            combined_axes_names = '_'.join(axes_names[:-1])
+            combined_labels = ['_'.join([str(vlabels[i][j]) for i in range(len(vlabels))])
+                               for j in range(len(vlabels[0]))]
+            axes = [la.Axis(combined_labels, combined_axes_names)]
+        # last axis
+        axes += [la.Axis(hlabels, axes_names[-1])]
+        return la.LArray(raw_data, axes)
+
+    def move_axis(self, data, bg_value, old_index, new_index):
+        assert isinstance(data, la.LArray)
+        new_axes = data.axes.copy()
+        new_axes.insert(new_index, new_axes.pop(new_axes[old_index]))
```

Comment:
I think this should be a method on AxisCollection

Reply:
See https://github.com/liam2/larray/issues/564